File: cpp/snake.cpp
#include <iostream>
#include <deque>
#include <cstdlib>
#include <ctime>
#include <vector>

#ifdef _WIN32
    #include <conio.h>
    #include <windows.h>
    #define CLEAR system("cls")
#else
    #include <termios.h>
    #include <unistd.h>
    #include <fcntl.h>
    char getch() {
        struct termios oldt, newt;
        char ch;
        tcgetattr(STDIN_FILENO, &oldt);
        newt = oldt;
        newt.c_lflag &= ~(ICANON | ECHO);
        tcsetattr(STDIN_FILENO, TCSANOW, &newt);
        ch = getchar();
        tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
        return ch;
    }
    #define CLEAR system("clear")
#endif

using namespace std;

const int WIDTH = 40;
const int HEIGHT = 20;
deque<pair<int,int>> snake;
pair<int,int> food;
int score = 0;
char dir = 'd';

void setup() {
    snake.push_front({HEIGHT/2, WIDTH/2});
    snake.push_back({HEIGHT/2, WIDTH/2-1});
    snake.push_back({HEIGHT/2, WIDTH/2-2});
    srand(time(0));
    food = {rand() % HEIGHT, rand() % WIDTH};
    score = 0;
    dir = 'd';
}

void draw() {
    CLEAR;
    for (int i = 0; i < HEIGHT+2; i++) cout << '#';
    cout << endl;
    for (int i = 0; i < HEIGHT; i++) {
        for (int j = 0; j < WIDTH; j++) {
            if (j == 0) cout << '#';
            bool isSnake = false;
            for (auto &seg : snake) if (seg.first == i && seg.second == j) isSnake = true;
            if (isSnake) cout << 'O';
            else if (food.first == i && food.second == j) cout << 'F';
            else cout << ' ';
            if (j == WIDTH-1) cout << '#';
        }
        cout << endl;
    }
    for (int i = 0; i < HEIGHT+2; i++) cout << '#';
    cout << "\nScore: " << score << "   Use WASD. Press q to quit.\n";
}

void input() {
    if (_kbhit() || true) {
        char c;
        #ifdef _WIN32
            c = tolower(getch());
        #else
            c = tolower(getch());
        #endif
        if (c == 'a' && dir != 'd') dir = 'a';
        else if (c == 'd' && dir != 'a') dir = 'd';
        else if (c == 'w' && dir != 's') dir = 'w';
        else if (c == 's' && dir != 'w') dir = 's';
        else if (c == 'q') exit(0);
    }
}

void logic() {
    pair<int,int> newHead = snake.front();
    if (dir == 'a') newHead.second--;
    else if (dir == 'd') newHead.second++;
    else if (dir == 'w') newHead.first--;
    else if (dir == 's') newHead.first++;

    if (newHead.first < 0 || newHead.first >= HEIGHT || newHead.second < 0 || newHead.second >= WIDTH) {
        cout << "\nGAME OVER!\n";
        exit(0);
    }
    for (auto &seg : snake) if (seg == newHead) { cout << "\nGAME OVER!\n"; exit(0); }

    snake.push_front(newHead);
    if (newHead == food) {
        score++;
        food = {rand() % HEIGHT, rand() % WIDTH};
    } else {
        snake.pop_back();
    }
}

int main() {
    setup();
    while (true) {
        draw();
        input();
        logic();
        #ifdef _WIN32
            Sleep(100);
        #else
            usleep(100000);
        #endif
    }
    return 0;
}

File: go/snake.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"time"
)

type Point struct {
	X, Y int
}

var snake []Point
var food Point
var dir string
var score int
var width, height = 40, 20

func clear() {
	cmd := exec.Command("clear")
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func initGame() {
	snake = []Point{{height / 2, width / 2}, {height / 2, width/2 - 1}, {height / 2, width/2 - 2}}
	dir = "RIGHT"
	score = 0
	rand.Seed(time.Now().UnixNano())
	food = Point{rand.Intn(height), rand.Intn(width)}
}

func draw() {
	clear()
	for i := 0; i < height+2; i++ {
		fmt.Print("#")
	}
	fmt.Println()
	for i := 0; i < height; i++ {
		for j := 0; j < width; j++ {
			if j == 0 {
				fmt.Print("#")
			}
			isSnake := false
			for _, s := range snake {
				if s.X == i && s.Y == j {
					isSnake = true
					break
				}
			}
			if isSnake {
				fmt.Print("O")
			} else if food.X == i && food.Y == j {
				fmt.Print("F")
			} else {
				fmt.Print(" ")
			}
			if j == width-1 {
				fmt.Print("#")
			}
		}
		fmt.Println()
	}
	for i := 0; i < height+2; i++ {
		fmt.Print("#")
	}
	fmt.Printf("\nScore: %d | Use WASD. Press q to quit.\n", score)
}

func readInput() {
	reader := bufio.NewReader(os.Stdin)
	ch, _, _ := reader.ReadRune()
	switch ch {
	case 'a', 'A':
		if dir != "RIGHT" {
			dir = "LEFT"
		}
	case 'd', 'D':
		if dir != "LEFT" {
			dir = "RIGHT"
		}
	case 'w', 'W':
		if dir != "DOWN" {
			dir = "UP"
		}
	case 's', 'S':
		if dir != "UP" {
			dir = "DOWN"
		}
	case 'q', 'Q':
		os.Exit(0)
	}
}

func update() {
	head := snake[0]
	var newHead Point
	switch dir {
	case "RIGHT":
		newHead = Point{head.X, head.Y + 1}
	case "LEFT":
		newHead = Point{head.X, head.Y - 1}
	case "UP":
		newHead = Point{head.X - 1, head.Y}
	case "DOWN":
		newHead = Point{head.X + 1, head.Y}
	}
	if newHead.X < 0 || newHead.X >= height || newHead.Y < 0 || newHead.Y >= width {
		fmt.Println("\nGAME OVER!")
		os.Exit(0)
	}
	for _, seg := range snake {
		if seg == newHead {
			fmt.Println("\nGAME OVER!")
			os.Exit(0)
		}
	}
	snake = append([]Point{newHead}, snake...)
	if newHead == food {
		score++
		food = Point{rand.Intn(height), rand.Intn(width)}
	} else {
		snake = snake[:len(snake)-1]
	}
}

func main() {
	initGame()
	go func() {
		for {
			readInput()
		}
	}()
	for {
		draw()
		update()
		time.Sleep(100 * time.Millisecond)
	}
}
